### ISSUE TITLE
Ambiguous Type Resolution?

### ISSUE THREAD

**Issue by MatthewChang** [rmosolgo/graphql-ruby#491]
Forgive my ignorance if this is not a true issue.

It appears as this implementation of graph ql departs from the standard by having a single type resolution function defined on the schema. This in itself is not a problem, but the context passed into that function seems to not be the context at the evaluating node. I.e. 
```ruby
GraphQL::Schema.define do
  query QueryType
  resolve_type -> (object,ctx) {
    puts ctx.irep_node
  }
end
```
is always resulting in nil in my schema (ctx.ast_node raises a NoMethodError). I would guess that this is the context of the schema.

Given that, if I were to have two union types which operate on the same kind of resolved object, such as unions for query and create operations, it is ambiguous simply from the type of the object which type to resolve to. This could be circumvented by embedding some meta information in the object as to which kind of operation it is a part of, but that seems like a patently bad design pattern. 
Have I missed something with the proper way to specify type resolution? Thanks for your help!

**Comment by rmosolgo** [comment]
You raise a good point! Yes, that's the query-level context. It contains any keys that you passed as `context:`, but it doesn't have the field-level values. It seems like it should, huh? 

Here's the bad line of code: 

https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/execution/execute.rb#L178

It calls through to `query`, which provides its own context. But that method signature could be improved to `query.resolve_type(value, field_ctx)`, that way it would work as you expect! 

Perhaps another way to improve it would be to inject the composite type which is being resolved, eg: 

```ruby
resolve_type ->(obj, type, ctx) { 
  # 
}
```

Where `type` is a union or interface type. (If `ctx` were improved as described above, I think you could get this value via `ctx.irep_node.owner_type`.)

BTW, the current API was designed to _prevent_ the same feature requested here 😆  It was previously possible to resolve the _same_ objects to _different_ types -- but when that happened accidentally, it resulted in nasty bugs! From your example, I see how that ability can be _useful_ in the right situation!

**Comment by MatthewChang** [comment]
Ah I see. Thanks for the prompt and thorough response! I haven't quite needed this functionality yet (and designing around it does seem like the preferable option) but may make a PR if I end up using it. Cheers!

**Comment by khamusa** [comment]
I'd really like to see this implemented too. I think I could give it a try. Do you think `query.resolve_type(value, field_ctx, query_ctx)` would make sense? I'm aware that `query_ctx` is also accessible as `field_ctx.query.context`, but providing the query context directly as an argument tends to reduce dependencies and make tests more approachable.

What do you think ?

@rmosolgo could you give an example of the type of nasty bugs that used to occur when it was possible for unions/interfaces to define their own `resolve_type`?

**Comment by rmosolgo** [comment]
What's the use case for _both_ `query_ctx` and `field_ctx`? Why is it helpful to have `query_ctx` when you have `field_ctx`? (I expect everything from `query_ctx` to be available directly from `field_ctx`.)

Regarding type resolution bugs, that was my diagnosis of the issues surrounding https://github.com/rmosolgo/graphql-ruby/pull/209 although it was never confirmed. Typecasting of various unions and interfaces _could_ return different object types for the same object, resulting in nondeterministic query execution. For example, `NodeInterface.resolve_type(some_obj)` might return a different type than `SearchResultUnion.resolve_type(some_obj)`!

**Comment by khamusa** [comment]
@rmosolgo I guess field_ctx is really enough. I'll see if I can come up with a PR still this week.

**Comment by rmosolgo** [comment]
This is addressed in 1.6.5, check the changelog for new options :)
